url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/org/jgraph/graph/DefaultGraphModel.java

```diff
@@ -0,0 +1,2121 @@
+					}
+				} else {
+					if (parent instanceof DefaultMutableTreeNode
+							&& child instanceof MutableTreeNode) {
+						((DefaultMutableTreeNode) parent)
+								.add((MutableTreeNode) child);
+					}
+				}
+
+				if (rootsSet == null) {
+					rootsSet = new HashSet(roots);
+				}
+				boolean isRoot = rootsSet.contains(child);
+				if (parent == null && !isRoot) {
+					rootsSet.add(child);
+					roots.add(child);
+				}
+				else if (parent != null && isRoot) {
+					if (rootsToBeRemoved == null) {
+						rootsToBeRemoved = new HashSet();
+					}
+					rootsSet.remove(child);
+					rootsToBeRemoved.add(child);
+				}
+			}
+			if (rootsToBeRemoved != null && rootsToBeRemoved.size() > 0) {
+				// If any roots have been removed, reform the roots
+				// lists appropriately, keeping the order the same
+				int newRootsSize = roots.size() - rootsToBeRemoved.size();
+				if (newRootsSize < 8) {
+					newRootsSize = 8;
+				}
+				List newRoots = new ArrayList(newRootsSize);
+				Iterator iter = roots.iterator();
+				while (iter.hasNext()) {
+					Object cell = iter.next();
+					if (!rootsToBeRemoved.contains(cell)) {
+						newRoots.add(cell);
+					}
+				}
+				roots = newRoots;
+			}
+			return undo;
+		}
+		return null;
+	}
+
+	/**
+	 * Applies <code>attributes</code> to the cells specified as keys. Returns
+	 * the <code>attributes</code> to undo the change.
+	 */
+	protected Map handleAttributes(Map attributes) {
+		if (attributes != null) {
+			Hashtable undo = new Hashtable(attributes.size());
+			Iterator it = attributes.entrySet().iterator();
+			while (it.hasNext()) {
+				Map.Entry entry = (Map.Entry) it.next();
+				Object cell = entry.getKey();
+				Map deltaNew = (Map) entry.getValue();
+				// System.out.println("deltaNew="+deltaNew);
+				// System.out.println("stateOld="+getAttributes(cell));
+				// Handle New Values
+				Map deltaOld = null;
+				AttributeMap attr = getAttributes(cell);
+				if (attr != null) {
+					deltaOld = attr.applyMap(deltaNew);
+					// System.out.println("stateNew="+getAttributes(cell));
+					// System.out.println("deltaOld="+deltaOld);
+					undo.put(cell, deltaOld);
+				} else {
+					// Make room for the value
+					deltaOld = new Hashtable(2);
+				}
+				// Handle new values
+				Object newValue = deltaNew.get(GraphConstants.VALUE);
+				if (newValue != null) {
+					Object oldValue = valueForCellChanged(cell, newValue);
+					if (oldValue != null)
+						GraphConstants.setValue(deltaOld, oldValue);
+					// TODO: Userobject of null is probably invalid
+					else
+						GraphConstants.setRemoveAttributes(deltaOld,
+								new Object[] { GraphConstants.VALUE });
+				} else {
+					// Special case to handle removal of value attribute
+					Object[] remove = GraphConstants
+							.getRemoveAttributes(deltaNew);
+					if (remove != null && remove.length > 0) {
+						for (int i = 0; i < remove.length; i++) {
+							if (remove[i] == GraphConstants.VALUE) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.